opening terminal with a command function and tests added

ADIthaker opened the pull request:
## Description

Opens `cmd` for windows and gnome-terminal for linux based systems with user-specified command and gives the user complete access to the terminal window as it is detached from the program.

## Changes

The changes apply to the ergo/electron folder which has a platform folder, this folder exports functions based on the OS
currently exports only the `openTerm.js` file

## Issue

> Closes #21 and  #2 
 
## Impacted Area

Created a backend utility function

## Steps to test

Steps needed to reproduce the scenario from this change to validate if the pull request solve this issue

-  `npm run test`
- while testing on windows make sure that you will have to delete the file manually because the `cmd` spawned keeps it open 
in its process, so u will see error `EBUSY` can't unlink file, resource is busy/ being used

ADIthaker committed:
opening terminal with a command function and tests added

opens platform dependant terminals and run user-specified commands in them

ADIthaker committed:
added changes to event listening setup from Ipc renderer patch pull request

ADIthaker committed:
added changes to event listening setup from Ipc renderer patch pull request
--missed a file

ADIthaker committed:
Merge branch 'new-open-term' of https://github.com/ADIthaker/ergo into new-open-term

ADIthaker committed:
fixed a commit error

YJDoc2 commented on test/openTerm.js:
```diff
@@ -0,0 +1,35 @@
+const { assert } = require("chai");
+const { after } = require("mocha");
+const log = require("electron-log");
+const openTermLinux = require("../app/electron/platform/linux/openTerm");
+const openTermWin = require("../app/electron/platform/windows/openTerm");
```
As index of platform exports correct one, directly import that. In case that works incorrectly, then it will also be tested as well.

YJDoc2 commented on test/openTerm.js:
```diff
@@ -0,0 +1,35 @@
+
+after(() => {
+  unlink("./test.txt", (err) => {
+    if (err) log.error(err);
+    else log.info("Test completed and file deleted");
```
Always use {} for if-else, even when not required, usually improves readability. Also, don't log this here, as this will be logged in electron log file, which should not contain test logs

ADIthaker replied:
im not allowed by eslint to use the default console because we have added this [rule](https://eslint.org/docs/rules/no-console). I researched a bit but found some [code](https://stackoverflow.com/questions/31759367/using-console-log-in-electron-app#answer-48923571) doing the same.

YJDoc2 commented on test/openTerm.js:
```diff
@@ -0,0 +1,35 @@
+    else log.info("Test completed and file deleted");
+  });
+});
+if (process.platform === "linux") {
+  describe("Open Terminal", () => {
+    it("Should open terminal and create file [test.txt]", async () => {
+      const arrCommandType = ["touch", "./test.txt"];
+      await openTermLinux(arrCommandType);
+      const isFile = existsSync("./test.txt");
+      assert(isFile, true);
+    });
+  });
+} else if (process.platform === "win32") {
+  describe("Open Terminal", () => {
+    it("Should open terminal and create file [test.txt]", async () => {
+      const arrCommandType = ["echo", ".>", "test.txt"];
+      await openTermWin(arrCommandType);
+      const isFile = existsSync("./test.txt");
+      assert(isFile, true);
+    });
+  });
```
Rather than putting describe in condition, as most of part in them is common, except arrCommandType (after importing from index) declare that and assign value to it in if-else; for eg:
let arrCommandType;
if(process.platform=== 'linux'){
arrCommandType = ...
}else...

YJDoc2 commented on app/electron/preload.js:
```diff
@@ -15,8 +14,7 @@ contextBridge.exposeInMainWorld("api", {
   contextMenu: ContextMenu.preloadBindings(ipcRenderer),
-  //  send: name => sendHello(name),
-  //  get: ()=> getCollege(),
+  api: events.preloadBindings(ipcRenderer, fs),
```
Check if we can give this name better than "api", as this object is already going to be in api, so the call from frontend will be window.api.api.something()

YJDoc2 commented on app/electron/platform/windows/openTerm.js:
```diff
@@ -0,0 +1,24 @@
+  return cmd.concat(arrCommand);
+};
+
+module.exports = (arrCommand) => {
+  return new Promise((resolve, reject) => {
+    const executableArray = insertCommandWin(arrCommand);
+    const child = spawn("cmd", executableArray, {
+      detached: true,
+      stdio: ["ignore", "ignore", "ignore"],
+    });
+    child.on("error", (err) => {
+      reject(err);
+    });
+    child.on("exit", () => {
+      resolve(0);
+    });
+    child.unref();
+  });
```
Great job :+1:

YJDoc2 commented on app/electron/platform/windows/openTerm.js:
```diff
@@ -0,0 +1,24 @@
+const { spawn } = require("child_process");
+
+const cmd = ["/k", "start", "cmd.exe", "/k"];
```
Why are there two /k ?

YJDoc2 commented on app/electron/platform/linux/openTerm.js:
```diff
@@ -0,0 +1,23 @@
+const { exec } = require("child_process");
+
+const gnomeTerminal = "gnome-terminal --tab --active -- bash -c ";
+const xTerm = "xterm -fa 'Monospace' -fs 12 -e /bin/bash -l -c ";
+// append command to xTerm and then append ';exec bash';
+// append command to gnomeTerminal and then append '; exec bash'
+const insertCommandLinux = (arrCommand, type) => {
+  const command = arrCommand.join(" ");
+  if (type === "xterm") {
+    return `${`${xTerm}'${command}`};/bin/bash'`;
+  }
+  return `${`${gnomeTerminal}'${command}`}; exec bash'`;
+};
+module.exports = (arrCommand, type = "gnomeTerminal") => {
+  return new Promise((resolve, reject) => {
+    const executableString = insertCommandLinux(arrCommand, type);
+    exec(executableString, (err) => {
+      if (err) {
+        reject(err);
+      } else resolve(0);
+    });
+  });
```
Great work :+1:
Try if we can open user's preferred shell rather than hard-coding bash.
doing 'echo $SHELL' will give path to user's shell (/bin/bash), maybe that can be used instead of hard-coding.
Try to implement this, as users may have some other shell such as fish or zsh or some other.

ADIthaker replied:
good catch thanks

YJDoc2 reviewed (commented):
Overall great work :+1:, just try to make the few suggested changes.

ADIthaker committed:
added a few suggested changes

opens default shell
formatting in conditionals
testing exports from /app/electron/platform

YJDoc2 reviewed (approved):
:+1:
